### PR TITLE
Fix garbage collection bugs

### PR DESCRIPTION
**Pull request by cptbtptpbcptdtptp**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/galacean/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### Other information:

### CODE REVIEW

**Review by jtabibito — COMMENTED** [review]
+1

File: packages/core/src/RenderPipeline/RenderContext.ts [hunk 1]
```diff
@@ -61,4 +61,8 @@ export class RenderContext {
     projectionParams.set(flipProjection ? -1 : 1, virtualCamera.nearClipPlane, virtualCamera.farClipPlane, 0);
     shaderData.setVector4(RenderContext._cameraProjectionProperty, projectionParams);
   }
+
+  garbageCollection(): void {
```
Comment (GuoLei1990) [inline packages/core/src/RenderPipeline/RenderContext.ts:65]: Should not do this

File: packages/core/src/shader/ShaderProgramPool.ts [hunk 2]
```diff
@@ -64,6 +71,20 @@ export class ShaderProgramPool {
     this._cacheMap = Object.create(null);
   }
 
+  private _recursiveGC(hierarchy: number, cacheMap: Tree<ShaderProgram>): void {
```
Comment (GuoLei1990) [inline packages/core/src/shader/ShaderProgramPool.ts:74]: There seems to be no need to do this

File: packages/core/src/RenderPipeline/RenderContext.ts (hunk 1 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/RenderPipeline/RenderContext.ts:65]: Done.

**Pull request merged**